test/system/700-play.bats

```diff
@@ -871,93 +897,62 @@ spec:
     - touch /tmp/healthy && sleep 100
     livenessProbe:
       exec:
+        # /tmp/healthy will exist in healthy container
+        # /tmp/unhealthy will never exist, and will thus
+        # cause healthcheck failure
         command:
         - cat
-        - /tmp/healthy
+        - /tmp/$want
       initialDelaySeconds: 3
       failureThreshold: 1
       periodSeconds: 1
-" > $fname
-
-    run_podman kube play $fname
-    ctrName="$podname-$ctrname"
-
-    # Keep checking status. For the first 2 seconds it must be 'starting'
-    t0=$SECONDS
-    while [[ $SECONDS -le $((t0 + 2)) ]]; do
-        run_podman inspect $ctrName --format "1-{{.State.Health.Status}}"
-        assert "$output" == "1-starting" "Health.Status at $((SECONDS - t0))"
-        sleep 0.5
-    done
+EOF
 
-    # After 3 seconds it may take another second to go healthy. Wait.
-    t0=$SECONDS
-    while [[ $SECONDS -le $((t0 + 3)) ]]; do
-        run_podman inspect $ctrName --format "2-{{.State.Health.Status}}"
-        if [[ "$output" = "2-healthy" ]]; then
-            break;
+        run_podman kube play $fname
+        ctrName="$podname-$ctrname"
+
+        # Collect status every half-second until it goes into the desired state.
+        local i=1
+        local full_log=""
+        while [[ $i -lt 15 ]]; do
+            run_podman inspect $ctrName --format "$i-{{.State.Health.Status}}"
+            full_log+=" $output"
+            if [[ "$output" =~ "-$want" ]]; then
+                break
+            fi
+            sleep 0.5
+            i=$((i+1))
+        done
+
+        assert "$full_log" =~ "-$want\$" \
+               "Container got to '$want'"
+        assert "$full_log" =~ "-starting.*-$want" \
+               "Container went from starting to $want"
+
+        if [[ $want == "healthy" ]]; then
+            dontwant="unhealthy"
+        else
+            dontwant="healthy"
         fi
-        sleep 0.5
-    done
-    assert $output == "2-healthy" "After 3 seconds"
-
-    run_podman kube down $fname
-}
-
-@test "podman kube play healthcheck should wait initialDelaySeconds before updating status (unhealthy)" {
-    podname="liveness-exec-$(safename)"
-    ctrname="liveness-ctr-$(safename)"
+        assert "$full_log" !~ "-$dontwant" \
+               "Container never goes $dontwant"
+
+        # GAH! Save ten seconds, but in a horrible way.
+        #   - 'kube down' does not have a -t0 option.
+        #   - Using 'top' in the container, instead of 'sleep 100', results
+        #     in very weird failures. Seriously weird.
+        #   - 'stop -t0', every once in a while on parallel runs on my
+        #     laptop (never yet in CI), barfs with 'container is running or
+        #     paused, refusing to clean up, container state improper'
+        # Here's hoping that this will silence the flakes.
+        run_podman '?' stop -t0 $ctrName
```

Comment:
Do you have a log for this? What exactly is not working with top?

Reply:
I have lots of logs for "container state improper", is there anything useful that can be gleaned from them?

`top`: no, I can't find those logs any more. I will try to reproduce.

Reply:
I mean podman stop should be idempotent so it should never error with container state improper here so this is likely something that needs to be fixed in podman

Reply:
Isn't it the inverse in this check  `"Container never goes $dontwant"` So the `healthy` case failed?
The log doesn't make much sense 5-unhealthy for it to then go back to starting, these seems like something wrong with health checks.
But yeah sleep vs top no idea...

Reply:
Oops yeah, this is the `healthy` test failing. I only see that when I change `sleep 100` to `top` or `top -b`. I keep trying to understand it and keep failing to.

Reply:
Another one:
```
#|     FAIL: Container got to 'healthy'
#| expected: =~ -healthy\$
#|   actual:     1-starting 2-starting 3-starting 4-starting 5-starting 6-starting 7-starting 8-starting 9-starting 10-starting 11-starting 12-starting 13-starting 14-starting
```
I'll try bumping up the timeout, but 7.5 seconds does seem like enough for a 3-second `initialDelay`...

Reply:
nope:
```
#|     FAIL: Container got to 'unhealthy'
#| expected: =~ -unhealthy\$
#|   actual:     1-starting 2-starting 3-starting 4-starting 5-starting 6-starting 7-starting 8-starting 9-starting 10-starting 11-starting 12-starting 13-starting 14-starting 15-starting 16-starting 17-starting 18-starting 19-starting 20-starting 21-starting 22-starting 23-starting 24-starting
```
Giving up. I'm just monkeying without actually learning anything useful.

Reply:
How easy is this to reproduce? I can try to instrument some podman code to see if I can figure out what is going on.

Reply:
Here's a log of the `stop` issue, from Sep 5:
```
...
# [11:57:04.641519074] # bin/podman kube play /tmp/podman_bats.NjbFo3/play_kube_unhealthy_7iPZ8Q.yaml
# [11:57:05.565840635] Pod:
# 5f7428be75f02b786e801f19c9bb494011c8e12d70a6ae29e809f3463a6fe16f
# Container:
# 8ae0ccceac07f06a879da59b02771f7699abf390003b440c18f98947678dbee7

...

# [11:57:09.139196762] # bin/podman inspect liveness-exec-t383-vn2b6p1n-unhealthy-liveness-ctr-t383-vn2b6p1n-unhealthy --format 5-{{.State.Health.Status}}
# [11:57:09.296619275] 5-unhealthy
#
# [11:57:09.322312911] # bin/podman stop -t0 liveness-exec-t383-vn2b6p1n-unhealthy-liveness-ctr-t383-vn2b6p1n-unhealthy
# [11:57:10.034340238] Error: container 8ae0ccceac07f06a879da59b02771f7699abf390003b440c18f98947678dbee7 is running or paused, refusing to clean up: container state improper
```

Reply:
...and, this just in, it finally just triggered on a local run. No need to post log, because it looks exactly the same as above modulo container names and shas. Anyhow, the point is, `podman stop` is barfing in the way you say should not happen. But it's rare.